### PR TITLE
Lintcheck: Add URL to lint emission place in diff

### PR DESCRIPTION
**Pull request by xFrednet**
This PR adds links to the emission code in our lintcheck CI. When reviewing changes, I would like to be able to see the bigger context, which isn't always included in the lint message. This PR adds a nice link to the lintcheck diff that allows for simple investigation of the code in question.

At first, I wanted to simply use the doc.rs links and call it a day, but then I figured out that some crates might have their source files remapped. Cargo was the crate that made me notice this. As a response, I made the link configurable. (See https://github.com/rust-lang/docs.rs/issues/2551 for a detailed explanation and possible solution to remove this workaround in the future.)

It's probably easiest to review the individual commits. The last one is just a dummy to showcase the change.

 
[:framed_picture: rendered :framed_picture:](https://github.com/rust-lang/rust-clippy/actions/runs/9960834924?pr=13104)

---

r? @Alexendoo 

changelog: none

---

That's it, I hope that everyone who's reading this has a beautiful day :D

### CODE REVIEW

**Comment by xFrednet** [comment]
It looks like the cargo link is sadly still incorrect :thinking:

**Comment by xFrednet** [comment]
The cargo links work now :tada: 

Some links are currently not 100% correct as the versions we use for lintcheck are so old that they used a previous version of docs.rs link structure. I considered fixing this, but decided to hold of on it, until I update our default and CI test set (Which is on my todo list for next week)

File: clippy_lints/src/casts/unnecessary_cast.rs
Comment (flip1995) [inline clippy_lints/src/casts/unnecessary_cast.rs:55]: :eyes: Don't forget to remove this again :)

Comment (xFrednet) [reply clippy_lints/src/casts/unnecessary_cast.rs:55]: Or we could create an RFC suggesting that all lint messages should from now on include ASCII emojis :P

(Yep, I'll undo these changes, the commit also has a big "do not merge" in the name)

**Comment by Alexendoo** [comment]
A few times I've forgotten if I'm reading the `Added` or `Removed` sections because they look the same, this would add a convenient place to put a plus/minus or similar, what do you think?

**Comment by xFrednet** [comment]
I like the idea. A simple `+` or `-` would be made into a list by GH's Markdown. I now used the title as a prefix, which constructs messages like this:

> Added `clippy::cool_lint` at https://cataas.com/cat
> Removed `clippy::cool_lint` at https://cataas.com/cat
> Changed `clippy::cool_lint` at https://cataas.com/cat

File: lintcheck/src/json.rs
Comment (Alexendoo) [inline lintcheck/src/json.rs:13]: Could remove these now since the `file_link` would cover it well enough

Comment (xFrednet) [reply lintcheck/src/json.rs:13]: The byte position is still used in the `key()` function on line 19. We could replace it with the link, but I think having it separately would make more sense. While all links currently contain line labels, there is no guarantee that they will for manually configured lints. And links also don't contain the column, which is included in this field.

Comment (Alexendoo) [reply lintcheck/src/json.rs:13]: Not critical though, either way it's ready for r+ after ditching the test commit

Comment (xFrednet) [reply lintcheck/src/json.rs:13]: > Removing information from the key has a benefit - it would mean tweaks to a span's start column/end line/end column show as a change rather than

The tweaked span would still result in a `added` and `removed` message, since the lint message displays the span and is also used as part of the span. I think here it's safer to display them separately and leave it to the reviewer to merge them.

Comment (Alexendoo) [reply lintcheck/src/json.rs:13]: The message isn't part of the key

File: lintcheck/src/input.rs
Comment (Alexendoo) [inline lintcheck/src/input.rs:15]: Ah one last thing - typo in `DEFAULT`, missed that until now 😄

Comment (xFrednet) [reply lintcheck/src/input.rs:15]: *That was totally intentional, and just a check if you actually check the changes /s :P*

**Review by Alexendoo — APPROVED** [review]

**Comment by Alexendoo** [comment]
🐰 

@bors r+

**Pull request merged**
